### ISSUE TITLE
[bug] Some AutoLoads are removed when WAT is enabled

### ISSUE THREAD

**Issue by TheDuckCow** [AlexDarigan/WAT#314]
When WAT is enabled at the same time as another addon which defines an AutoLoad (at the very least, when the [Firebase addon](https://github.com/GodotNuts/GodotFirebase) is enabled), the non-WAT AutoLoad is removed from the save file.

More specifically said: This is problematic as I require a specific order of AutoLoads, and because somehow WAT is removing this Firebase AutoLoad. While the Firebase addon automatically adds it back since it sees it is missing, it just puts it at the end, when I need it to be ordered first. 

## Are you sure this isn't a Firebase addon issue?

Mostly. See the related issue here for more context: https://github.com/GodotNuts/GodotFirebase/issues/270

Firebase works fine on its own, only when WAT is enabled does this issue occur. That's why I think it may affect any addon that defines an AutoLoad, but I don't know of another addon to test that assumption. 

## Reproducible example

See the uploaded zip file and follow this sequence, zip contains all latest versions of all addons involved including WAT:
[autoload-issue.zip](https://github.com/AlexDarigan/WAT/files/7585994/autoload-issue.zip)

0. Inspect the `godot.project` file, notice that the `Firebase="*res://addons/godot-firebase/firebase/firebase.tscn"` line exists under the AutoLoad section.
1. Load the project.
2. Notice how there are no errors on open
3. Go into Project Settings, addons: See that WAT is disabled. Enable it
4. Save the project
5. Inspect the `godot.project` file, notice there are no changes (yet)
4. Quit the project/godot
5. Now when you inspect the `godot.project` file, notice that the `Firebase="*res://addons/godot-firebase/firebase/firebase.tscn"` line was somehow removed.
6. Re-open godot, and notice that the Firebase autoload was moved to the end (as the Firebase addon recognized it was missing, and so it re-added it.. but at the end of course). Also notice all the errors on startup.

Right now, the net result is that I can't run all tests, especially those built on top of Firebase, and the project always opens with excessive errors. This behavior also occurs even when I'm exporting games by the command line, as after each export command completes it resets the autoload (removes the Firebase one).

It is worth calling out that in my 'real' project, I don't have any other AutoLoads defined by addons, so I have no idea if this somehow specifically clashes with the GodotNut Firebase addon, or if it would interfere with any addon that defines an AutoLoad Singleton. But the issue goes away if I disable WAT altogether, pointing to some issue here. It does NOT appear to affect AutoLoads defined by my project outside the addons folder, as you would find in the sequence of steps above.

**Comment by TheDuckCow** [comment]
Also, not to complicate this more, but this also uncovered another somewhat concerning behavior:

I have a command line script like so:
```
echo "Running tests"
"$GODOT_EXEC" addons/WAT/cli.tscn run=all
```

When I'm running it such that the `Firebase="*res://addons/godot-firebase/firebase/firebase.tscn"` line is presently saved to the `godot.project` file, I get 4 test outputs:

```
-------RESULTS-------
Took 6 second(s)
0 Tests Crashed
2 / 4 Tests Passed
-------RESULTS-------
```

But if I open and close the project, or even just run an export (even if the export is run from command line... which is actually part of this same script itself), this same `"$GODOT_EXEC" addons/WAT/cli.tscn run=all` call  claims there was 1 fewer test, along with the slew of printouts related to my first (due to the firebase autoload being loaded later than other autoloads which depend on it):

```
-------RESULTS-------
Took 2 second(s)
0 Tests Crashed
2 / 3 Tests Passed
-------RESULTS-------
```

**The thing that is concerning, is that the CLI WAT test doesn't recognize it failed to load/run this test.** It doesn't even show up under the "Tests Crashed" section. So I might not even realize some tests files are being entirely skipped if they are fatally broken. Not sure if there's a way around that, realize this may be more a restriction of how the addon works.

**Comment by RailKill** [comment]
This is reproducible on my machine using Godot 3.4, Windows 10. However, this is a Godot engine bug. WAT doesn't even deal with Autoloads in any way, it's very self-contained. This is further cemented by the fact that commenting out ```remove_autoload_singleton()``` on line 8 of the [GodotFirebase's plugin.gd](https://github.com/GodotNuts/GodotFirebase/blob/godotfirebase-v4.2/addons/godot-firebase/plugin.gd) resolves the issue. So @TheDuckCow can use this as a workaround for now.

The thing is, ```_exit_tree()``` of Godot EditorPlugin is always called **when the editor is closed**, and I've confirmed that this is the case by adding code to create a file in ```_exit_tree()``` - the file is created every time the Godot editor is closed. So why does ```remove_autoload_singleton()``` **not** remove the Firebase autoload on exit? That is the bug with Godot engine here. It is only when WAT is enabled, ```remove_autoload_singleton()``` begins to work as intended and removes Firebase autoload on exit.

This issue should thus be addressed back to @WolfgangSenff. To address https://github.com/GodotNuts/GodotFirebase/issues/270, I think instead of creating and removing the singleton in plugin.gd ```_enter_tree()``` and ```_exit_tree()```, perhaps it should only be done during plugin enable and disable by overriding [```enable_plugin()```](https://docs.godotengine.org/en/stable/classes/class_editorplugin.html#class-editorplugin-method-enable-plugin) and [```disable_plugin()```](https://docs.godotengine.org/en/stable/classes/class_editorplugin.html#class-editorplugin-method-disable-plugin)? It doesn't make sense for the Firebase singleton to be recreated every time the Godot engine is restarted. I've opened a pull request for your review if this simple change is acceptable. Feel free to reject/edit if unfit: https://github.com/GodotNuts/GodotFirebase/pull/273

**Comment by TheDuckCow** [comment]
Thanks a bunch for going the extra mile opening a PR in the other repo, and finding a solution for me here. Also appreciate the quick work addressing the point of failing tests not being picked up.

Ignoring the autoload/firebase thing for now since the other PR address this, I went ahead and tried out your change from the `improve-test-detection` branch. I realize there are also other changes in that branch, so could be a few things going on, but I'm happy to see another bug fixed that I hadn't yet commented about (where dropdowns fail to populate after first running a script when one of them has a fatal error).

The only thing I do notice is I get an unexpected number of "total tests" via command line. I see 8 being referenced, suspiciously the same as the number of tests from the first file:
<img width="1146" alt="tests in ui" src="https://user-images.githubusercontent.com/2958461/143719713-edd2fb6f-e182-4dd5-9c8a-ba28f9960d08.png">

And command line run all results via `"$GODOT_EXEC" addons/WAT/cli.tscn run=all` is:
```
-------RESULTS-------
Took 2 second(s)
0 Test Scripts Crashed
3 / 8 Test Scripts Passed
-------RESULTS-------
```
From prior behavior, I expect the number to be 4 instead of 8, regarding the total number of test files; if it's intended to now represent the total number of individual tests (which I would definitely say would be preferred!), the total numbers don't seem to add up. In the UI screenshot, I would expect the numbers to say (in order left to right):

`| (8/8) res://... |(2/2) res://... |(4/4) res://... | (1/3) res://...  |`

And thus I'd expect the Test Scripts Passed to be more like 15/17 instead of 3/8 (I really have no idea where the 3 comes from... maybe 8 is coming from the first test total somehow?).

**Comment by TheDuckCow** [comment]
Hm one fast follow, I just ran my command line script again, and now am seeing

-------RESULTS-------
Took 7 second(s)
0 Test Scripts Crashed
2 / 4 Test Scripts Passed
-------RESULTS-------

I tried rerunning a couple times on command line, and getting the same. 2 /4 (vs being 3/8). Seems odd, but there is one flaky tests in one of the other files which could be a culprit.

I just ran one more time after closing the interface (so only command line is running), and I'm once more seeing:

-------RESULTS-------
Took 0 second(s)
0 Test Scripts Crashed
2 / 8 Test Scripts Passed
-------RESULTS-------

Running again, I get 3 / 8 (again, possibly due to one flaky test I've got). Just figure I'd share!

**Comment by RailKill** [comment]
@TheDuckCow The total number of test should maintain the current behavior, so it seems that it is a bug in my fork. Hmm, it's very strange that the number of tests being picked up will change if you just close the Godot interface. I've pushed some new commits to my fork, you can give it a test to see if this problem still persists: https://github.com/RailKill/WAT/tree/improve-test-detection.

Let me know if there's any output that catches your eye. Whenever a test fails, it should show the test script name, followed by its reason. If not, try scrolling up to see if you find any errors output by Godot itself which can help narrow down the issue for me. Thanks!

**Comment by TheDuckCow** [comment]
Just tried it again using the latest from that branch., seems to be the same.

I then tried again after applying your [patch here](https://github.com/GodotNuts/GodotFirebase/pull/273/files) to my local firebase plugin. Now I'm consistently getting the expected 2/4 and printouts now of the tests failing (which, the failing tests make sense and are expected). So I wonder if there's still some other complicating factor with the other plugin...

--

Something I do note is when I open the project, or when running tests from the command line, I get these prinouts in the Output window or terminal console (I get x7 of these in printouts):
```
SCRIPT ERROR: Trying to assign value of type 'SceneTree' to a variable of type 'Node'.
          at: _get_test_script (res://addons/WAT/filesystem/filesystem.gd:94)
```

To further test/validate, I just now replaced my WAT install with the code from this most [recent release](https://github.com/AlexDarigan/WAT/releases/tag/v6.0.0-Mono). Now,  I'm not seeing the above errors.

If I revert back to the latest from the `improve-test-detection` branch, the above error comes back, but consistently prints out 2/4 tests (which is the expected behavior) via command line.

**Comment by TheDuckCow** [comment]
`so the latest commits implemented a RegEx check to look at the source code of broken scripts,`

Just want to say, that's super going the extra mile by doing that, amazing! I'll try to give it a try soon, but apologies if it takes me awhile as I'm getting caught up with some other items. But your explanation and approach makes sense to me.

**Comment by AlexDarigan** [comment]
Please check if this is solved from the commit 790b30094931b5da1d6913c377174da12d0c3908

**Comment by AlexDarigan** [comment]
I'm going to assume this is resolved based on your prior discussions. I'll reopen if the problem persists.
